### PR TITLE
fix: Autofix timeout bug

### PR DESCRIPTION
**Pull request by BBerabi**
### Description

Autofix Diffs flow (web view flow) was implemented slightly differently than the code action flow. The diff flow treated an empty list of suggestions as an error and invalid response. This is not the correct behavior because the AI Fix pipeline might run successfully and there can be no valid fixes found. 

The code was polling until the suggestion's length was not 0. This causes the returned successful response with empty suggestions to be ignored and the code aggressively keeps polling until the timeout hits. The timeout takes 2 minutes whereas the result is available after several seconds. The timeout also hides the real issue from the user, making the user think something unexpected happened whereas we could not compute any valid fix. 

This PR corrects the flow for autofix diffs by

- adding status into the return value similar to how it is done for code actions
- corrects the logic in the polling function

### Checklist

- [x] Tests added and all succeed
- [x] Linted
- [x] README.md updated, if user-facing
- [x] License file updated, if new 3rd-party dependency is introduced

I uploaded a video to my drive showing the incorrect behavior on the IDE.
https://drive.google.com/file/d/1BlHmbkW5WEiSdPQKzmjoU-5IspK20PoZ/vi

and here the corrected behaviour which immediately shows user the result of 0 fixes

https://github.com/user-attachments/assets/61b3c417-cfd0-4822-90b9-4592866e9ffe

### CODE REVIEW

**Comment by BBerabi** [comment]
I think the failing test is not related to my changes but because of an RCE vulnerability detected by Snyk Code. Is anyone already looking at it?

**Review by ShawkyZ — COMMENTED** [review]
Thank you Berkay
It looks good to me besides some nitpicks.

Let's pair on it on Monday. I would like to revisit / understand some stuff here, like why do we have the AutofixStatus struct and if we can simplify/refactor it a bit.

File: domain/ide/command/code_fix_diffs.go [hunk 1]
```diff
@@ -83,6 +83,10 @@ func (cmd *codeFixDiffs) Execute(ctx context.Context) (any, error) {
 	}
 
 	suggestions, err := cmd.codeScanner.GetAutoFixDiffs(ctx, folderPath, relPath, issue)
+	if err == nil && len(suggestions) == 0 {
+		logger.Info().Msgf("Autofix run successfully but there were no good fixes")
+		return suggestions, err
```
[the remark this answers is not in the document]
Comment (BBerabi) [reply domain/ide/command/code_fix_diffs.go:88]: done

File: domain/ide/command/code_fix_diffs.go [hunk 2]
```diff
@@ -83,6 +83,10 @@ func (cmd *codeFixDiffs) Execute(ctx context.Context) (any, error) {
 	}
 
 	suggestions, err := cmd.codeScanner.GetAutoFixDiffs(ctx, folderPath, relPath, issue)
+	if err == nil && len(suggestions) == 0 {
+		logger.Info().Msgf("Autofix run successfully but there were no good fixes")
```
[the remark this answers is not in the document]
Comment (BBerabi) [reply domain/ide/command/code_fix_diffs.go:87]: yes, makes sense, fixed

File: infrastructure/code/autofix.go [hunk 3]
```diff
@@ -51,19 +55,36 @@ func (s *SnykCodeHTTPClient) GetAutoFixDiffs(ctx context.Context, baseDir string
...
+	logger.Debug().Msgf("Status: %s", response.Status)
+
+	if response.Status == failed.message {
+		logger.Err(err).Str("responseStatus", response.Status).Msg("autofix failed")
+		return nil, failed, err
 	}
 
-	return response.toUnifiedDiffSuggestions(baseDir, options.filePath), err
+	if response.Status == "" {
+		s.c.Logger().Err(err).Str("responseStatus", response.Status).Msg("unknown response status (empty)")
```
[the remark this answers is not in the document]
Comment (BBerabi) [reply infrastructure/code/autofix.go:77]: This handles the unexpected case where the connection breaks or something unexpected happens and response.Status stays empty. This should never be the case if the call is returned. 

To be honest, I took this from the code action flow. There we do the same logic. 

I converted the s.c.Logger to logger.

File: infrastructure/code/autofix.go [hunk 4]
```diff
@@ -41,29 +40,48 @@ func (a AutofixUnifiedDiffSuggestion) GetUnifiedDiffForFile(filePath string) str
 	return a.UnifiedDiffsPerFile[filePath]
 }
 
-func (s *SnykCodeHTTPClient) GetAutofixDiffs(ctx context.Context, baseDir string, options AutofixOptions) (unifiedDiffSuggestions []AutofixUnifiedDiffSuggestion, err error) {
+func (s *SnykCodeHTTPClient) GetAutofixDiffs(ctx context.Context, baseDir string, options AutofixOptions) (
+	unifiedDiffSuggestions []AutofixUnifiedDiffSuggestion,
+	status AutofixStatus,
+	err error,
+) {
```
Comment (bastiandoetsch) [inline infrastructure/code/autofix.go:47]: on github it looks like it's weirdly formatted. just nitpicking, though.

File: infrastructure/code/autofix.go [hunk 5]
```diff
@@ -41,29 +40,48 @@ func (a AutofixUnifiedDiffSuggestion) GetUnifiedDiffForFile(filePath string) str
...
-		return unifiedDiffSuggestions, err
+	response, err := s.RunAutofix(span.Context(), options)
+	if err != nil {
+		return nil, failed, err
+	}
+
+	logger.Debug().Msgf("Status: %s", response.Status)
+
+	if response.Status == failed.message {
+		logger.Error().Str("responseStatus", response.Status).Msg("autofix failed")
```
Comment (bastiandoetsch) [inline infrastructure/code/autofix.go:70]: This probably should be an info level or warn level, as it's not a technical error, but a business logic outcome, right?

File: infrastructure/code/autofix.go [hunk 6]
```diff
@@ -41,29 +40,48 @@ func (a AutofixUnifiedDiffSuggestion) GetUnifiedDiffForFile(filePath string) str
...
+		logger.Error().Str("responseStatus", response.Status).Msg("unknown response status (empty)")
+		return nil, failed, errors.New("Unknown response status (empty)")
 	}
 
-	return response.toUnifiedDiffSuggestions(baseDir, options.filePath), err
+	if response.Status != completeStatus {
+		return nil, status, nil
+	}
+
+	logger.Debug().Msgf("Running toUnifiedDiffSuggestions")
```
Comment (bastiandoetsch) [inline infrastructure/code/autofix.go:83]: do we need this?

File: infrastructure/code/autofix.go [hunk 7]
```diff
@@ -109,13 +127,17 @@ func (sc *Scanner) GetAutofixDiffs(
...
-				logger.Err(err).Msg("Error getting autofix suggestions")
-				return nil, err
-			}
-			if len(suggestions) > 0 {
-				return suggestions, nil
+			suggestions, fixStatus, autofixErr := codeClient.GetAutofixDiffs(span.Context(), baseDir, options)
+			if autofixErr != nil {
+				logger.Err(autofixErr).Msg("Error getting autofix suggestions")
+				return nil, autofixErr
+			} else if fixStatus.message == completeStatus {
```
Comment (bastiandoetsch) [inline infrastructure/code/autofix.go:134]: Isn't there the case missing when `autofixErr == nil` and `fixStatus.message != completeStatus`? I'd at least leave a comment for that case that it's gonna loop - which it probably should not do if the status is already `FAILED`.

**Review by bastiandoetsch — APPROVED** [review]
Some nitpicks - please decide yourself what to update :). Thanks for the PR, Berkay!

File: infrastructure/code/autofix.go (hunk 6 above)
Comment (BBerabi) [reply infrastructure/code/autofix.go:83]: no, we do not, good catch, removed

File: infrastructure/code/autofix.go (hunk 5 above)
Comment (BBerabi) [reply infrastructure/code/autofix.go:70]: Good point, Antoine in our team is currently working on providing more granular error propagation on the backend side. I took this from another place in the codebase. I keep it as ERROR for now and we will revisit this in the downstream services once we add granular errors to the backend and sast-autofix-api.

File: infrastructure/code/autofix.go (hunk 7 above)
Comment (BBerabi) [reply infrastructure/code/autofix.go:134]: I have added a comment for that case :+1: 

if the status is already failed, then I made sure that we return an Error. So, in that case it is guaranteed to go into the branch `autofixErr != nil`

**Pull request merged**